### PR TITLE
Add Hopper config for the routemaster worker

### PR DESCRIPTION
**Pull request by muscar**
Should the task be marked as essential?

### CODE REVIEW

**Review by bencord0 — APPROVED** [review]
Looks harmless enough

**Comment by marcinwyszynski** [comment]
You're not using `decrypt_env`? Or `tini`?

**Comment by marcinwyszynski** [comment]
It's not the task that's marked essential but the container. It's the only container in the task, so you probably don't need to mark it as essential because if that container dies, the task dies anyway. So I guess it depends on whether you want to be explicit about that or not.

**Comment by muscar** [comment]
@marcinwyszynski Fair point re `decrypt_env`. I'll create a circle config to build & push worker images which use `decrypt_env`.

Re making the container essential, that was my reasoning as well. Thanks for confirming.

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,56 +1,151 @@
...
+  steps:
+  - setup_remote_docker:
+      reusable: true
+      version: 17.05.0-ce
+
+  - run:
+      name: Push to Image Repository
+      command: |
+        `print_env ${TARGET}`
+        push_image_to_ecr \
```
Comment (marcinwyszynski) [inline .circleci/config.yml]: ☣️ ☣️ ☣️ Here be dragons. ☣️ ☣️ ☣️

Please export the artifact between steps because they may be ran on different machines.

Comment (muscar) [reply .circleci/config.yml]: Right. I'll do that. Thanks.

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,56 +1,171 @@
...
+  - run:
+      name: Import image
+      command: docker load --input workspace/${CIRCLE_SHA1}.tar
+
+  - run:
+      name: Push image to repository
+      command: |
+        `print_env ${TARGET}`
+        push_image_to_ecr \
+          --image-name "${CIRCLE_PROJECT_REPONAME}-${ROUTEMASTER_PROCESS}" \
```
Comment (marcinwyszynski) [inline .circleci/config.yml]: Not sure about that - why do do we need a separate image for each process?

Comment (muscar) [reply .circleci/config.yml]: That's a good question. The idea was that we decide which process to run in the Dockerfile. If we can do that in Hopper, e.g. choose the command in the config file, then we don't need one image per process.

Comment (marcinwyszynski) [reply .circleci/config.yml]: You can pass it dynamically to the environment through command: `sh -c 'ENV=val blah'`. We could make it even easier by allowing folks to specify per-service environment inside Hopper config.

File: .hopper/config.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+version: '1'
+
+services:
+  worker:
+    containerDefinitions:
+      worker:
+        cpu: 256
+        memory: 1024
+        entryPoint: ["dumb-init", "--", "decrypt_env"]
```
Comment (marcinwyszynski) [inline .hopper/config.yml]: Please use the built-in (`tini`).

File: .hopper/config.yml [hunk 4]
```diff
@@ -6,5 +6,7 @@ services:
       worker:
         cpu: 256
         memory: 1024
-        entryPoint: ["dumb-init", "--", "decrypt_env"]
-        command: ["sh", "-c", "foreman start worker"]
+        linuxParameters:
+          initProcessEnabled: true
+        entryPoint: ["decrypt_env"]
```
Comment (marcinwyszynski) [inline .hopper/config.yml]: Any reason not to put it in `Dockerfile`?

File: .hopper/config.yml [hunk 5]
```diff
@@ -6,5 +6,7 @@ services:
...
+        command: ["foreman", "start", "worker"]
```
Comment (marcinwyszynski) [inline .hopper/config.yml]: You can also just do `foreman start worker` - Hopper will tokenize it for you.

File: Dockerfile [hunk 6]
```diff
@@ -24,8 +24,5 @@ USER $ARG_USER
 
 ARG ARG_PORT=3000
 ENV PORT=$ARG_PORT
-EXPOSE $PORT
 
-ARG ARG_PROCESS=web
-ENV PROCESS=$ARG_PROCESS
-CMD ["sh", "-c", "foreman start ${PROCESS}"]
+EXPOSE $PORT
```
Comment (marcinwyszynski) [inline Dockerfile]: No need for that.

File: Dockerfile [hunk 7]
```diff
@@ -24,8 +24,5 @@ USER $ARG_USER
 
 ARG ARG_PORT=3000
 ENV PORT=$ARG_PORT
```
Comment (marcinwyszynski) [inline Dockerfile]: Where do you import `decrypt_env`?

**Review by marcinwyszynski — APPROVED** [review]
image

**Comment by muscar** [comment]
Thanks for the helpful feedback, Marcin :).

**Pull request closed without merging**